### DISCUSSION TITLE
Crystal shapes, orientation problem

### DISCUSSION THREAD

**Discussion by Isa-EBSD** [mtex-toolbox/mtex#2146] [Ask Anything]
I want to display crystal shapes on top of grains.
I based the code off the mtex Crystal Shapes page; here are 3 variations.
mtex 5.11.2
```matlab
%%Grain reconstruction:
  cS = crystalShape.hex(ebsd('Quartz-new').CS)
  [grains, ebsd.grainId] = calcGrains(ebsd('Quartz-new'),'alpha',2.2,'angle',10*degree);
  grains = smooth(grains,5);
  isBig = grains.grainSize>50;

%% Option 1:
  cSGrains = grains(isBig).meanOrientation * cS * 0.2 * sqrt(grains(isBig).area); 
  figure
  plot(grains,grains.meanOrientation)
  hold on
  plot(grains(isBig).centroid + cSGrains)
  hold off

%% Option 2:
  figure
  plot(grains,grains.meanOrientation)
  hold on
  plot(grains(isBig),0.5*cS,'FaceColor',[0.8 0.8 0.8],'linewidth',1) 
  hold off

%% Option 3:
  figure
  plot(grains.boundary)
  hold on
  plot(grains, cS * 0.8,'linewidth', 1, 'colored')
  hold off
```
I compared the figures obtained with mtex with those generated in Aztec Crystal; using the selected option shown in image during import, the map with grain boundaries and LH PFs match perfectly, but not the orientation of the crystal shapes. 
![1139A2-Crystal shapes](https://github.com/mtex-toolbox/mtex/assets/170981953/eadfa841-f053-49ff-8699-72afe94e5fa1)

Any ideas on how to correct the orientation, so that everything matches?

**Comment by zmichels** [comment]
I'm not sure, and I'm only guessing here... I wonder if perhaps the plotting convention (with +Y to south) is affecting CS-drawing coordinates, such that that even if you've got the pole figures and grain structures aligned to match Aztec, MTEX uses a specific directionality for the the CS coordinate reference faces/shapes, too... which maybe is causing the 3D shapes to render inverted(?). IDK.

In the past, to avoid some different alignment issues... I used a combination like the following:

```matlab

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','outOfPlane');

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','ctf','convertEuler2SpatialReferenceFrame');

% rotations
ebsd = rotate(ebsd,rotation('axis',yvector,'angle',180*degree));
ebsd = rotate(ebsd,rotation('axis',zvector,'angle',180*degree),'keepEuler');

```

**Reply by Isa-EBSD** [reply]
Thank you very much for your answer!

I used the suggested rotations and now the map, crystal shapes, and pole figures (for the entire mapped area and for individual grains) match between Aztec and MTEX. 

The only discrepancy is in the Euler angles, with the 1st one being off by 180°. Is that a problem or is the dataset now considered properly aligned? I will be doing CVA analysis next and I want to ensure that everything is correct and aligned. Thank you!

![1139A2-Crystal shapes2](https://github.com/mtex-toolbox/mtex/assets/170981953/1bcf2b58-df4a-48fe-9e85-7838027411d8)

**Reply by zmichels** [reply]
- [ ] If I understand correctly, you have achieved an alignment between your orientation data and spatial reference frame in MTEX that matches the alignment in the acquisition system and AZtec. This is the most important aspect, and as you perhaps appreciate, it is not trivial. Part of the issue stems from the potentially unique geometry of an acquisition system... in which the +XYZ stage/chamber coordinates, vs the +XYZ imaging coordinate-system of the camera collecting Kikuchi patterns projected on the phosphor screen/sensor, vs the coordinate system/direction/alignment of any sensor used for mapping base image (like SE, BSE, FSE)... and which of these end up recorded by the mapping hard+software system configuration. In most cases... the Euler angles recorded by the acquisition system are relative to a view from the sensor's perspective, whereas the spatial/map coordinates are either related to the stage or to some scanning/imaging directionality... so they are different... and the proprietary software is designed to know all this and display the data relative to the maps and hemispheres you display. However, such details of the system are not transmitted/recorded in any metadata in EBSD files. When asking MTEX to plot any orientation data relative to some spatial directions (eg, pole figure, IPF-Z map, crystal shapes, etc)... MTEX needs to know how to reconcile these differences. Some setups have a common/similar geometry which is why I mentioned the rotations I've tried. When [eventually] everything as-plotted in MTEX matches (probably/potentially requiring some rotations) how it looks/looked in the acquisition system... it is very likely/sensible that the Euler angles no longer match those in the original EBSD file.

**Comment by kilir** [comment]
Hi,
 also consider that 
a) in mtex,the upper hemisphere refers to the positive z-direction
b) x0,y0 you have in your Aztec pf are aligned to the x0,y0 of your Aztec map.
Hopw that helps.
Cheers, Rüdigee

**Comment by PennyWieser** [comment]
Hi, I have the same issue (using a new Symmetry S3 on a JEOL IT800). 
However, I found that the suggested rotations crashes Matlab every time when I go onto plot the pole figure, without fail, on every laptop + desktop I have tried it on (including  very high powered ones). 

```matlab
figure(2)
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'angle',10*degree,'boundary','tight');
small = grains.grainSize<1000 ;
ebsd(grains(small)).phase = 0;
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'angle',10*degree,'boundary','tight');

hQ = [Miller(1,1,-2,0,grains('q').CS,'uvw'),Miller(0,0,0,1,grains('q').CS,'uvw')];

plotPDF(grains('q').meanOrientation,hQ,'complete','MarkerSize',10)`
```
What does seem to work is this:
```matlab
ebsd = rotate(ebsd,rotation('axis',xvector,'angle',180*degree));
ebsd = rotate(ebsd,rotation('axis',zvector,'angle',180*degree),'keepXY');`
```

As with the above poster, the Euler angles no longer match in my case - the first number is 180 degrees higher in MTEX than AztecCrystal, other two are the same, but everything else seems good. 

Are these rotations equivalent? Any idea why one goes onto crash that pole figure code and the other doesnt? 

This is using a ctf, if I export the exact same file as h5oina, I can get the crystal shape, pole figure, and map the right way around, but the euler angles are different again - should I just not worry about the euler angles? 
e.g. crystal: 178, 51, 90
ctf flip : 359, 51, 90
h50ina: 181, 129, 270
* note the small 1-2 degree differences are selecting the wrong pixel, its the big differences i'm worried about


![image](https://github.com/mtex-toolbox/mtex/assets/65668673/74ee4b32-f091-4a7a-a76c-1658f5b4d8d3)
![image](https://github.com/mtex-toolbox/mtex/assets/65668673/4554b5e1-6bbf-480d-8374-6fac9ec15682)
![image](https://github.com/mtex-toolbox/mtex/assets/65668673/323e5443-afd0-4bc8-a523-090e43fe95ce)

**Reply by ralfHielscher** [reply]
1. You should not worry about specific values of Euler angles. Euler angles are with respect to the specimen coordinate system. Unfortunately, most vendors use for the Euler angle a different specimen coordinate system as for the x,y coordinates of the EBSD map. This is a very confusing situation. MTEX always interprets Euler angles and x, y coordinates as with respect to the same reference frame. In order the achieve this it offers some functionality - see https://mtex-toolbox.github.io/EBSDReferenceFrame.html.
2. Not only is oxford using different reference frames for Euler angles and x, y coordinates - those coordinate systems are different between the different software product the company sells and the different formats it exports their data into. Therefore, import might work out of the box for one format and fail for the other one. 
3. You need to be careful setting up the right reference frames in MTEX. This is a work that needs to be done once - and allows you to import all subsequent data sets without any issue.

Ralf.

**Reply by Isa-EBSD** [reply]
Hi @PennyWieser,
I'm still working on figuring this all out! 
A colleague of mine suggested running a reference sample with a known orientation for the a, m and pi faces and c-axis. In theory, this would allow tracking the orientation of the data all the way through from thin section to Aztec to mtex. (We have some here at the university that we used with an automated fabric analyser but otherwise it should be possible to make a thin section from a large idiomorphic quartz grain.)
This might be the only way to know without any doubt whether the different codes result in equivalent rotations and to ensure that the orientation is preserved correctly. It would also answer the Euler angle question!
Once I've run the reference samples, I'll update this thread with my findings.

**Comment by PennyWieser** [comment]
@zmichels how did you get AztecCrystal to overlay the crystal shapes?
![image](https://github.com/user-attachments/assets/88be25f8-f3b3-49bd-a8ac-81dac4e71155)

**Reply by zmichels** [reply]
I think someone else posted that -- I do not have an Oxford system, and unfortunately I'm not sure of the specific procedure for plotting the crystal shapes in the aztec software.

**Reply by PennyWieser** [reply]
Oops sorry @Isa-EBSD

**Reply by Isa-EBSD** [reply]
Hi @PennyWieser ,
To display the shapes in Aztec Crystal, go to Map display settings>grain display>model: unit cell.
I originally found that through the Oxford training videos, Understanding Orientation Maps in EBSD (https://www.youtube.com/watch?v=1V3mlsILcGk&t=122s)
